client/state-db/src/lib.rs

```diff
@@ -481,14 +510,125 @@ pub struct StateDb<BlockHash: Hash, Key: Hash> {
 
 impl<BlockHash: Hash + MallocSizeOf, Key: Hash + MallocSizeOf> StateDb<BlockHash, Key> {
 	/// Creates a new instance. Does not expect any metadata in the database.
-	pub fn new<D: MetaDb>(
+	#[deprecated]
+	pub fn new<D: MetaDbRead>(
 		mode: PruningMode,
 		ref_counting: bool,
 		db: &D,
 	) -> Result<StateDb<BlockHash, Key>, Error<D::Error>> {
 		Ok(StateDb { db: RwLock::new(StateDbSync::new(mode, ref_counting, db)?) })
 	}
 
+	pub fn open<D>(
+		db: &mut D,
+		requested_mode: Option<PruningMode>,
+		ref_counting: bool,
+		should_init: bool,
+	) -> Result<Self, Error<Either<<D as MetaDbRead>::Error, <D as MetaDbWrite>::Error>>>
+	where
+		D: MetaDbRead,
+		D: MetaDbWrite,
+	{
+		let meta_key_mode = to_meta_key(PRUNING_MODE, &());
+		let meta_key_mode_arg = to_meta_key(PRUNING_MODE_ARG, &());
+
+		let stored_mode = db.get_meta(&meta_key_mode).map_err(Either::Left).map_err(Error::Db)?;
+		let stored_mode_arg =
+			db.get_meta(&meta_key_mode_arg).map_err(Either::Left).map_err(Error::Db)?;
+
+		let stored_mode = {
+			match stored_mode.as_ref().map(AsRef::as_ref) {
+				None => None,
+				Some(PRUNING_MODE_ARCHIVE) => Some(PruningMode::ArchiveAll),
+				Some(PRUNING_MODE_ARCHIVE_CANON) => Some(PruningMode::ArchiveCanonical),
+				Some(PRUNING_MODE_CONSTRAINED) =>
+					if let Some(mode_arg) = stored_mode_arg {
+						let mut decode_input = mode_arg.as_ref();
+						let (max_blocks, max_mem) = <PruningModeConstraintSerialized as codec::Decode>::decode(&mut decode_input)
+							.unwrap_or_else(|_|{
+								log::warn!("Failed to decode the value of meta PRUNING_MODE_ARG. Resorting to the default constraints");
+								Default::default()
+							});
```

Comment:
Since the `--validator` no longer implies `--pruning archive`, the omitted `--pruning <mode>` argument now will always yield the default mode — equivalent of `--pruning 256`.

But the old nodes after the fix should work as they used to work (without requiring to change their startup arguments).

That effectively brings some new behaviour: when `--pruning` is omitted but the storage has previously been initialised — use the mode that was previously set for the pruning.

If we decide to save only, whether the last mode was `archive` or `prune` (without the numeric arguments to the prune), then the whole fix would give the behaviour that I find a bit illogical: the storage carries the information of the last choice between `archive` and `prune`, but if you decide to restart the node with omitting the explicit pruning argument — it will fall back to the 256.

Reply:
> but if you decide to restart the node with omitting the explicit pruning argument — it will fall back to the 256.

This is already true for the current implementation and I don't see any problems with it. If you don't require any special pruning window, you just run with the default window that is fine enough. If you now decide to switch, you restart with a different cli flag. Next time you may decide again differently. However, this is an active choice. No real need to store this in the db.

The reason why `prune` or `archive` is stored in the db. You can not go from `prune` to `archive`. You can go from `archive` to `prune`, but to prevent fuck ups we don't allow this. However, when you change the pruning window, there is no real fuck up happening.

Reply:
Then I will remove saving the last preferred constraint